Upgrade to pip 19

cjolowicz opened the pull request:
This PR upgrades [pip](https://pypi.org/project/pip/) to version 19.x. This upgrade is the first step to provide support for `pyproject.toml`-based projects ([PEP 517](https://www.python.org/dev/peps/pep-0517/)), such as those managed by [poetry](https://poetry.eustace.io) and [flit](https://github.com/takluyver/flit).

- Upgrade to pip `19.2.3`, for projects using Python 2.7 and >= 3.5.
- Upgrade to pip `19.1.1`, for Python 3.4 projects. 

Note that pip 19.2 dropped Python 3.4 support, hence the version difference.

Patch the vendorized `pip-pop` package, which is used by the `pip-uninstall`, `gdal`, and `pylibmc` steps. It relies on internal pip modules changed by the update.

Closes #740
See #804

Follow-ups:

- #834 (Support pyproject.toml)
- #835 (Support poetry.lock)

---

CI is broken for PRs from another repository, see [this comment](https://github.com/heroku/heroku-buildpack-python/pull/832#issuecomment-524441204). Please check the corresponding PR in my fork to verify CI:

- cjolowicz/heroku-buildpack-python#13

cjolowicz commented:
pip just released 19.2.3.
https://pypi.org/project/pip/#history

David-OConnor commented:
Limitations on merging this?

cjolowicz commented:
Rebased onto master ([`v157`](https://github.com/heroku/heroku-buildpack-python/releases/tag/v157)).

CaseyFaist commented:
Thanks for the PR series @cjolowicz!

This looks good, but updating pip for pipenv users or requiring them to update without a heads up won't be a good experience (our version is old enough that they'll need to uninstall and reinstall pipenv locally to successfully update). If you can refactor this to stay pinned to current version for pipenv users only, I should be able to accept this (and the related project updates) 👍

CaseyFaist commented on vendor/pip-pop/pip-diff:
```diff
@@ -12,9 +12,24 @@ Options:
-from pip.req import parse_requirements
-from pip.index import PackageFinder
-from pip._vendor.requests import session
+
```
Can you say more about why these updates are required?

cjolowicz replied:
The `pip-diff` and `pip-grep` tools  import internal modules from pip. In pip >= 10, internal modules were moved under `pip._internal`, breaking the imports. This PR uses `try...except ImportError` to handle both import paths.

Also, the interface of the `PackageFinder` class from one of the internal pip modules changed. This PR provides a wrapper function to allow creating objects of this type using the old interface.

CaseyFaist replied:
Thanks for the context 💯

CaseyFaist commented:
also relevant: I'm going to extract the package manager detection parts from #816 as well, contributor willing. I can make sure it all plays nicely together after the fact, but all I'm looking for here is a separate `pipenv-pip-version` variable that I can pass to the pipenv step, and confirmation that the pip-diff updates should still work with the old version 👍

cjolowicz commented:
- Rebased onto master (af860be62afc610aa24e6a1556ab21279d3f5d0b)
- Added 8408e346396745e6bfd32ea38f1e3f5117278808 to address https://github.com/heroku/heroku-buildpack-python/pull/833#issuecomment-537758441, keeping pip at the current version for pipenv users

CaseyFaist commented:
@cjolowicz thanks for your work! Looking over it now, I'll add comments as I have them ❤️

helioseven commented:
What is the status of this PR? I see this as a high-priority fix, given that `manylinux2010` wheels do not work with pip 9.0.2 (as was mentioned in #804 ).

Currently, only option for building apps with such wheels seems to be forking this repo, modifying `PIP_UPDATE` in `/bin/compile`, and then setting buildpack for app to forked repo. Less than ideal, obviously.

cjolowicz commented:
Current pip release is now [19.3.1](https://github.com/pypa/pip/releases/tag/19.3.1) (October 17, 2019).

cjolowicz commented:
- Rebased onto master (93a620e6edacde750d32af6e3c8c3ad5e6800a40)
- Upgrade to pip 19.3.1 (40ec196d34a0774d6dc4519759407bcd2540cf4d)

cjolowicz commented:
@CaseyFaist : This PR is marked "changes requested", but I'm not sure which changes are requested.

helioseven commented:
Did it fail checks just because changelog.md wasn't modified? Do we just need to add an entry to said changelog for this PR?

cjolowicz commented:
@helioseven, the tests pass (with the exception of the changelog). See the PR in my fork:

- https://github.com/cjolowicz/heroku-buildpack-python/pull/13

CI in the upstream repository (here) does not work for external contributions.
